validator/validate_api.py

```diff
@@ -0,0 +1,312 @@
+        response,
+        remove_stopwords=remove_stopwords,
+        tag_numeric=tag_numeric,
+        correct_spelling=spelling_correction,
+        kill_nonwords=remove_nonwords,
+        spell_correction_max=spell_correction_limit,
+    )
+
+    # Initialize all feature counts to 0
+    # Then move through the feature list in order and count iff applicable
+    feature_count_dict = OrderedDict({key: 0 for key in feature_weight_dict.keys()})
+    feature_count_dict["intercept"] = 1
+
+    for word in response_words:
+        for key in feature_weight_dict.keys():
+            if feature_weight_dict[key]:
+                if word in feature_vocab_dict[key]:
+                    feature_count_dict[key] += 1
+                    break  # only count each word as first feature matched (key order matters!)
+
+    # Group the counts together and compute an inner product with the weights
+    vector = feature_count_dict.values()
+    WEIGHTS = feature_weight_dict.values()
+    inner_product = sum([v * w for v, w in zip(vector, WEIGHTS)])
+    valid = float(inner_product) > 0
+
+    return_dict = {
+        "response": response,
+        "remove_stopwords": remove_stopwords,
+        "tag_numeric": tag_numeric,
+        "spelling_correction_used": spelling_correction,
+        "num_spelling_correction": num_spelling_corrections,
+        "remove_nonwords": remove_nonwords,
+        "processed_response": " ".join(response_words),
+    }
+    return_dict.update(feature_count_dict)
+    return_dict["inner_product"] = inner_product
+    return_dict["valid"] = valid
+    return return_dict
+
+
+def validate_response(
+    response,
+    uid,
+    feature_weight_dict,
+    remove_stopwords=None,
+    tag_numeric=None,
+    spelling_correction=None,
+    remove_nonwords=None,
+    spell_correction_max=None,
+    lazy_math_mode=None,
+):
+    """Function to estimate validity given response, uid, and parser parameters"""
+
+    if remove_stopwords is None:
+        remove_stopwords = PARSER_DEFAULTS["remove_stopwords"]
+    if tag_numeric is None:
+        tag_numeric = PARSER_DEFAULTS["tag_numeric"]
+    if spelling_correction is None:
+        spelling_correction = PARSER_DEFAULTS["spelling_correction"]
+    if remove_nonwords is None:
+        remove_nonwords = PARSER_DEFAULTS["remove_nonwords"]
+    if spell_correction_max is None:
+        spell_correction_max = PARSER_DEFAULTS["spell_correction_max"]
+    if lazy_math_mode is None:
+        lazy_math_mode = PARSER_DEFAULTS["lazy_math_mode"]
+
+    # Try to get questions-specific vocab via uid (if not found, vocab will be empty)
+    # domain_vocab, innovation_vocab, has_numeric, uid_used, question_vocab,
+    #  mc_vocab = get_question_data(uid)
+    vocab_dict, uid_used, has_numeric = get_question_data(uid)
+
+    # Record the input of tag_numeric and then convert in the case of 'auto'
+    tag_numeric_input = tag_numeric
+    tag_numeric = tag_numeric or ((tag_numeric == "auto") and has_numeric)
```

Comment:
`"auto" or (True and False)` returns `"auto"` which doesn't seem to be the desired output.

Reply:
Ah, yup, this completely neuters has_numeric, doesn't it? Will fix.

Reply:
On further consideration, this PR didn't change this code (it just moved from one big `app.py` for everything to `validate_api.py`) currently applies the "lazy math" validity check to _everything_, so leads to false positives. Definitely file a bug, but I don't think it should block this merge.